src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -0,0 +1,366 @@
+        protected static void ThrowIncorrectCurrentIdException() =>
+            throw new InvalidOperationException(SR.InvalidOperation_IncorrectToken);
+
+        /// <summary>Describes states the operation can be in.</summary>
+        public enum States
+        {
+            /// <summary>The operation has been assigned an owner.  No one else can use it.</summary>
+            Owned = 0,
+            /// <summary>Completion has been reserved.  Only the reserver is allowed to complete it.</summary>
+            CompletionReserved = 1,
+            /// <summary>The operation has completed and has had its result or error stored.</summary>
+            CompletionSet = 2,
+            /// <summary>The operation's result/error has been retrieved.  It's available for reuse.</summary>
+            Released = 3
+        }
+    }
+
+    internal abstract class ResettableValueTaskSource<T> : ResettableValueTaskSource, IValueTaskSource, IValueTaskSource<T>
+    {
+        private volatile int _state = (int)States.Owned;
+        private T _result;
+        private ExceptionDispatchInfo _error;
+        private Action<object> _continuation;
+        private object _continuationState;
+        private object _schedulingContext;
+        private ExecutionContext _executionContext;
+        private short _currentId;
+
+        public ValueTask ValueTask => new ValueTask(this, _currentId);
+        public ValueTask<T> ValueTaskOfT => new ValueTask<T>(this, _currentId);
+
+        public bool RunContinutationsAsynchronously { get; protected set; }
+        public ValueTaskSourceStatus GetStatus(short token)
+        {
+            if (_currentId == token)
+            {
+                switch ((States)_state)
+                {
+                    case States.Owned:
+                    case States.CompletionReserved:
+                        return ValueTaskSourceStatus.Pending;
+
+                    case States.CompletionSet:
+                        return
+                            _error == null ? ValueTaskSourceStatus.Succeeded :
+                            _error.SourceException is OperationCanceledException ? ValueTaskSourceStatus.Canceled :
+                            ValueTaskSourceStatus.Faulted;
+                }
+            }
+
+            ThrowIncorrectCurrentIdException();
+            return default; // just to satisfy compiler
+        }
+
+        public bool IsCompleted => _state >= (int)States.CompletionSet;
+        public States UnsafeState { get => (States)_state; set => _state = (int)value; }
+
+        public T GetResult(short token)
+        {
+            if (_currentId != token)
+            {
+                ThrowIncorrectCurrentIdException();
+            }
+
+            if (!IsCompleted)
+            {
+                ThrowIncompleteOperationException();
+            }
+
+            ExceptionDispatchInfo error = _error;
+            T result = _result;
```

Comment:
Why do we store the _result in a local before returning?

Reply:
Because once _state is set to Released below, someone else can immediately start using this object, so we need the state captured locally.

Reply:
> Is it because after the _state changes, the value could change?

Right